src/lib/support/BufferReader.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "BufferReader.h"
+
+namespace chip {
+namespace Encoding {
+namespace LittleEndian {
+
+namespace {
+// These helper methods return void and put the value being read into an
+// outparam because that allows us to easily overload on the type of the
+// thing being read.
+void ReadHelper(const uint8_t *& p, uint8_t * dest)
+{
+    *dest = Read8(p);
+}
+void ReadHelper(const uint8_t *& p, uint16_t * dest)
+{
+    *dest = Read16(p);
+}
+void ReadHelper(const uint8_t *& p, uint32_t * dest)
+{
+    *dest = Read32(p);
+}
+void ReadHelper(const uint8_t *& p, uint64_t * dest)
+{
+    *dest = Read64(p);
+}
+} // anonymous namespace
+
+template <typename T>
+void Reader::Read(T * retval)
+{
+    static_assert(CHAR_BIT == 8, "Our various sizeof checks rely on bytes and octets being the same thing");
+
+    static constexpr size_t data_size = sizeof(T);
+
+    if (mAvailable < data_size)
+    {
+        mStatus = CHIP_ERROR_BUFFER_TOO_SMALL;
+        // Ensure that future reads all fail.
+        mAvailable = 0;
+        return;
+    }
+
+    ReadHelper(mReadPtr, retval);
+    mAvailable = static_cast<uint16_t>(mAvailable - data_size);
+}
+
+// Explicit Read instantiations for the data types we want to support.
+template void Reader::Read(uint8_t *);
```

Comment:
I noticed that our size actually increased. How about we place these in the header to allow for inlining instead? Does that make any difference in binary size?

Reply:
The size increased because we only have a few callsites of each function so far.  With inlining, we pay N bytes per callsite.  With them out of line, we pay 4N + M * (# of callsites) bytes, where M is the cost of the function call; M < N.

I did measurements after adding some more callsites (as part of the fix for <https://github.com/project-chip/connectedhomeip/issues/2168>) and at that point the out-of-line version is smaller than the inline one.